src/pybind/mgr/dashboard/models_proposal.md

```diff
@@ -0,0 +1,608 @@
+# Controller Model Proposal
+
+## Context and Motivation
+
+Currently the ceph-dashboard controller infrastructure allows to consume HTTP request data through
+Python method parameters.
+
+The HTTP request data can be of four kinds:
+* Path parameters
+* Query parameters
+* Body data (JSON encoded)
+* HTTP Headers (currently not consumed directly by the controller infrastructure)
+
+The several kinds of request data are binded to the endpoint Python method parameters by matching
+the parameter name against the key of the data (with the exception of path parameters, which are
+handled a bit different).
+
+This approach works quite well when the data we receive is of the form (key, value) where value is
+of some scalar type, i.e., contains a single value and not a complex data structure. But sometimes,
+we do receive much more complex data encoded in JSON of variable depth.
+
+For instance, consider the following examle of an HTTP body JSON structure:
+
+```json
+{
+    "name": "John Doe",
+    "contacts": {
+        "personal": "96111111",
+        "work": "91111111"
+    }
+}
+```
+
+and an endpoint method declared like the following:
+
+```python
+@Endpoint('POST')
+def create(self, name, contacts):
+  # ...
+```
+
+In the above example, when the HTTP request is issued, the controller infrastructure will parse the
+JSON data and will call the `create` method with the value `John Doe` for parameter `name`, and the
+value `{"personal": "96111111", "work": "91111111"}` for parameter `contacts`.
+This means that `contacts` variable will actually hold a python dict and there is no information
+in the code that can tell us that fact.
+Moreover, if we want to automatically generate documentation for our controllers, there is not
+enough information either.
+
+Besides the lack of information about the structure of data that is binded to each method parameter,
+there is also another feature lacking in the current infrastructure, which is a structured way
+of validating the request data.
+
+Many times the data received by our controllers are subject to restrictions of the domain, or even
+complex rules. Currently, such validation must be done manually in the endpoint method code, which
+leads to a lot of boilerplate code, and most of the times, to be completely inexistent.
+
+The purpose of this proposal is enhance the controller infrastructure with notion of data models
```

Comment:
The request-response conversation is the fundamental process that drives all communication on the web. 
Because of this, these models shouldn't be be called data models but "RequestModel / ResponseModel".
Examples:
Django: https://docs.djangoproject.com/en/2.2/ref/request-response/
Symfony: https://symfony.com/doc/current/introduction/http_fundamentals.html#symfony-request-object
Spring: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/RequestEntity.html